src/app/reporting/ReportScheduler.h

```diff
@@ -0,0 +1,160 @@
+/*
+ *
+ *    Copyright (c) 2023 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/ReadHandler.h>
+#include <lib/core/CHIPError.h>
+#include <lib/support/IntrusiveList.h>
+#include <system/SystemClock.h>
+
+namespace chip {
+namespace app {
+
+// Forward declaration of TestReportScheduler to allow it to be friend with ReportScheduler
+class TestReportScheduler;
+
+namespace reporting {
+
+using Timestamp = System::Clock::Timestamp;
+
+class ReportScheduler : public ReadHandler::Observer
+{
+public:
+    class TimerDelegate
+    {
+    public:
+        virtual ~TimerDelegate() {}
+        virtual CHIP_ERROR StartTimer(void * context, System::Clock::Timeout aTimeout) = 0;
```

Comment:
The behavior of this needs to be documented.  For example, if there is already a timer for this context, does this cancel the existing timer?  Or start a second one?  Or something else?

Reply:
This would depend on the implementation no? The ReportSchedulerImpl will cancel any timer existing for a given node when started so I can write it here. But this delegate can behave differently on different implementation (for instance, in my brand new SyncrhonizeReportSchedulerImpl, the delegate implemented in the test behaves differently than the one in the ReportSchedulerImpl, but that's another PR to come)

Reply:
I think that's fine, but that needs to be _documented_.  Because callers need to know whether calling StartTimer with the same context twice will call the timer callback once, or twice or whether they need to be ready for either one.

And if callers are not supposed to do that without canceling in between, then _that_ should be documented.